File: 0020-ValidParentheses/python/ValidParentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        close_to_open = {")": "(", "}": "{", "]": "["}
        stack = []

        for c in s:
            if c not in close_to_open:
                stack.append(c)
            else:
                if stack and stack[-1] == close_to_open[c]:
                    stack.pop()
                else:
                    return False
        return stack == []


if __name__ == "__main__":
    solution = Solution()

    s = "([])"
    result = solution.isValid(s)
    print(f"Input: s = \"{s}\"\nOutput: {result}")

File: 0020-ValidParentheses/cpp/ValidParentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> map_close = {{')', '('}, {'}', '{'}, {']', '['}};
        stack<char>st;
        for (auto e : s){
            if(!map_close.count(e)){
                st.push(e);
            }else{
                if (st.empty() || map_close[e] != st.top()){
                    return false;
                }
                st.pop();
            }
        }
        return st.empty();
    }
};